### ISSUE TITLE
Associative arrays with anonymous domains and no curly brackets

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#18887]
Should the compiler accept array declarations over anonymous associative domains without requiring curly brackets, like so?

```chapel
var A1: ["hi", "there"] real;
var A2: [1, 2, 3] real;
var A3: [A2] real;
```

In more detail:  Currently, the compiler accepts array declarations of the form:

```chapel
var A: [1..n, 1..n] real;
```

as a shorthand for:

```chapel
var A: [{1..n, 1..n}] real;
```

(that is, a 2D array over the anonymous domain `{1..n, 1..n}`), primarily as a convenience given that most languages don't have domains and so requiring the curly brackets seems like overkill in the rectangular cases.

It turns out that we also permit the curly brackets to be dropped for associative arrays, such that declarations of the form:

```chapel
var A: [i1, i2, i3] real;
var B: [i1] real;
```

(where i1, i2, i3 are of a similar type and not ranges) to mean an associative array over an anonymous domain, like:

```chapel
var A: [{i1, i2, i3}] real;
var B: [{i1}] real;
```

or:

```chapel
const D: domain(i1.type) = {i1, i2, i3};
var A: [D] real;
const D2: domain(i1.type) = {i1};
var B: [D2] real;
```

This surprised me and seems likely to cause confusion, particularly when the types of the indices are things like an array, a domain, a tuple of ranges or the like; or when the index is a singleton and the user thought they were specifying a domain.

To that end, I'm proposing that we require curly brackets for arrays over anonymous associative domains and curious whether anyone would object to this change.

**Comment by vasslitvinov** [comment]
I do not see this as particularly confusing.

While I'd rather keep it, this is a weak preference.

**Comment by aconsroe-hpe** [comment]
On the one hand, the rule for dropping curly brackets seems straightforward (assuming I'm understanding it correctly): 
```chapel
var A3: [xi ...] real; 
// becomes
var A3: [{xi ...}] real;
```

but they get different types (if I'm not mistaken)
```chapel
var A3: [x1 ... xN] real; 
// becomes
var A3: [{x1 ... xN}] real;
// {x1 ... xN} has type domain(N) if type(xi) is range(?)
// {x1 ... xN} has type domain(type(xi))
```

so I think I see why it could be confusing. Though I feel the confusion is more inherent in understanding a domain itself and less about the syntax.

I think I favor consistency so either requiring `{}` in all cases or none of them.

One potential gotcha that isn't an issue today (since its not supported b/c of no hash it looks like) is that you can't have a domain of type `domain(range)` (an associative domain where the keys are ranges -- maybe something mapping intervals to values). If `domain(range)` became a valid type then it makes even `{1..2}` ambiguous if you want `domain(range)` or `domain(1)`. This is tangent to the discussion though.

**Comment by mppf** [comment]
> I think I favor consistency so either requiring `{}` in all cases or none of them.

Even for `var A:[1..10, 1..20] int` ? This issue isn't proposing requiring curly braces for that. Are you?

**Comment by aconsroe-hpe** [comment]
> Even for var A:[1..10, 1..20] int ?

Yes.

>  This issue isn't proposing requiring curly braces for that. Are you?

Not proposing just sharing my thoughts

**Comment by e-kayrakli** [comment]
I support the proposal here.

> One potential gotcha that isn't an issue today (since its not supported b/c of no hash it looks like) is that you can't have a domain of type domain(range) (an associative domain where the keys are ranges -- maybe something mapping intervals to values). If domain(range) became a valid type then it makes even {1..2} ambiguous if you want domain(range) or domain(1). This is tangent to the discussion though.

I don't think this is tangential. I am not sure if there is any fundamental reason why we don't support `domain(range)`. So `var A: [1..10, 1..10] int` can be ambiguous to the user if they knew that we support the syntax in question. The fact that there is no `domain(range)` today does not clear that confusion to me. I must say that I am in general confused about associative domain literals for similar reasons.

The only annoyance I can think of that this proposal can bring is if you have a homogeneous tuple and you want to create an associative array over the values in that tuple. It turns into a parentheses galore quite easily:

```chapel
var A: [{(...myTuple)}] int;
```

This doesn't motivate me enough to argue against the proposal, though.

**Comment by bradcray** [comment]
> I am not sure if there is any fundamental reason why we don't support domain(range). 

I agree that we should support associative domains of ranges (`domain(range)`), and in fact, was working on enabling support for them (and associative domains of other types that support hashes) when I ran into this issue while updating tests.  I agree that `{1..n, 1..n}` is ambiguous once we have both and imagine that we'd break the ambiguity in favor of dense rectangular domains (where perhaps one could assign or cast or coerce from a dense rectangular domain to an associative domain of ranges to get the other).

I understand the desire for consistency that Andrew makes but am not wiling to go back to requiring people to write `var A: [{1..10}] real;` to get a 10-element array because I think it just raises barriers to using the language, so would stick with the status quo if consistency mattered to me over all else.

But Michael's arguments above are very much aligned with my thoughts—when I see uses of this, particularly with integer indices or indices specified as identifiers `var A: [i1, i2, i3] real;` it confuses me without seeming to provide a lot of value (whereas the value of permitting them to be dropped in the rectangular cases is essentially "because it's the closest equivalent to how other languages define arrays").

A conservative reason for deprecating or even removing the support is that it can always be added back later on if we decide we were being too skittish about it at present.  Whereas if we keep it through Chapel 2.0, we're arguably stuck with it.  My guess is that nobody will ever miss it / that most users don't even know about it.

**Comment by bradcray** [comment]
While working on this again yesterday, I realized that we may have worse ambiguity problems here than I had anticipated.  Specifically, there are two futures today that point out that an expression like `[1, 2, 3, 4](3)` fails because it's interpreted as the loop expression `[_ in {1, 2, 3, 4}] (3)` rather than what I think it's intended to be—namely, index into the 3rd element of this anonymous array.  I still need to dive into the parser to remind myself how these various cases are handled and what we could do, but it occurs to me there's a progression between the "obvious" array type declaration `[1..n, 1..n, 1..n] int` and `[1..n, 1..n, 1..n] typeExpr` and `[1..n, 1..n, 1..n] (typeExpr)` and `[r1, r2, r3] (typeExpr)` and `[r1, r2, r3] (i)` where the last case here looks indistinguishable from indexing into an anonymous array again—at least until we've resolved the types of the expressions, which is too late (in the current compiler), I think.

If I'm not mistaken so far, some possibilities for resolving this might include:
* only support curly brackets to be dropped from anonymous rectangular domains in such contexts when the `..` notation is visible in the source code (i.e., once I switch to `[r1, r2, r3]` I'd have to write it as `[{r1, r2, r3}]`
* only support the curly brackets being dropped in 1D array case (probably would affect a lot of existing code)
* something else?
